Basic virtual currency system

Meldanor opened the pull request:
Implements a basic approach to use a virtual currency (credits) inside SCION to reserve bandwidth. This pull requests is a dependency for a pull request in the scion-web project (https://github.com/netsec-ethz/scion-web/pull/105)

Content of the pull request:
- Every AS gets credits to create connections
- A new REST resource to view current connections and their costs
- Errors when a AS tries to create a connection without enough credits

For a more detailed view, please read this white paper (EDIT: 2nd reviewed version):
[SCION_VirtualCurrency_V2.pdf](https://github.com/netsec-ethz/scion-coord/files/1273269/SCION_VirtualCurrency_V2.pdf)


<!-- Reviewable:start -->
---
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/netsec-ethz/scion-coord/37)
<!-- Reviewable:end -->

ercanucan commented:
@meldanor I provided some initial feedback. As discussed with @hausheer, it would be good to have all the virtual currently related code in a separate module, and also with a flag or config variable to enable it. Also, it would be good if you can add some unit tests for the functions you have implement.

---

Review status: 0 of 4 files reviewed at latest revision, 12 unresolved discussions, some commit checks failed.

---

*[controllers/api/as.go, line 654 at r4](https://reviewable.io:443/reviews/netsec-ethz/scion-coord/37#-KvaaaOjlhwkkjSy2kjk:-KvaaaOkYzPyco2CnbuV:bkcag7t) ([raw file](https://github.com/netsec-ethz/scion-coord/blob/d48537844251649b0c92634eb6c7b59b5fb2e099/controllers/api/as.go#L654)):*
> ```Go
> 	fmt.Fprintln(w, string(b))
> }
> func (c *ASController) ListAsesConnectionsWithCredits(w http.ResponseWriter, r *http.Request) {
> ```

I would place all the functions related to virtual currency into a separate file, and call it from there.

---

*[controllers/api/as.go, line 695 at r4](https://reviewable.io:443/reviews/netsec-ethz/scion-coord/37#-Kva_BWKHUAAuowy_Ia9:-Kva_BWKHUAAuowy_IaA:bjovvub) ([raw file](https://github.com/netsec-ethz/scion-coord/blob/d48537844251649b0c92634eb6c7b59b5fb2e099/controllers/api/as.go#L695)):*
> ```Go
> 	fmt.Fprintln(w, string(b))
> 
> }
> ```

missing new line

---

*[models/as.go, line 65 at r4](https://reviewable.io:443/reviews/netsec-ethz/scion-coord/37#-KvaaMX1PsWgo-coEbQG:-KvaaMX1PsWgo-coEbQH:btbey7x) ([raw file](https://github.com/netsec-ethz/scion-coord/blob/d48537844251649b0c92634eb6c7b59b5fb2e099/models/as.go#L65)):*
> ```Go
> }
> 
> type ConnectionWithCredits struct {
> ```

Missing comments for the struct fields.

---

*[models/as.go, line 72 at r4](https://reviewable.io:443/reviews/netsec-ethz/scion-coord/37#-Kvaa9SBISU9mrCTaxIx:-Kvaa9SBISU9mrCTaxIy:b18f2nq) ([raw file](https://github.com/netsec-ethz/scion-coord/blob/d48537844251649b0c92634eb6c7b59b5fb2e099/models/as.go#L72)):*
> ```Go
> 	IsOutgoing bool
> 	Timestamp string
> ```

Unnecessary white-space. Please follow the coding style in the rest of this code-base as much as possible.

---

*[models/as.go, line 99 at r4](https://reviewable.io:443/reviews/netsec-ethz/scion-coord/37#-Kva_wE_aSfX9Q5AqvN3:-Kva_wE_aSfX9Q5AqvN4:bjq50wn) ([raw file](https://github.com/netsec-ethz/scion-coord/blob/d48537844251649b0c92634eb6c7b59b5fb2e099/models/as.go#L99)):*
> ```Go
> 
> 	// Ingoing ones (this AS get Credits for)
> 	var inGoings []ConnRequest
> ```

The variable `inGoings` does not really make sense. Do you mean incoming?

---

*[models/as.go, line 105 at r4](https://reviewable.io:443/reviews/netsec-ethz/scion-coord/37#-Kvac-kOhbkANjkAoMtM:-Kvac-kP1h10ljYSlbeD:b-zidegb) ([raw file](https://github.com/netsec-ethz/scion-coord/blob/d48537844251649b0c92634eb6c7b59b5fb2e099/models/as.go#L105)):*
> ```Go
> 	}
> 	for _, v := range inGoings {
> 		var tmp,_ = addr.IAFromString(v.RequestIA)
> ```

Would be better to have a variable name that's not `tmp`

---

*[models/as.go, line 140 at r4](https://reviewable.io:443/reviews/netsec-ethz/scion-coord/37#-Kva_phdbAHixoGtf4ap:-Kva_pheVXzSwhxDzg6W:b-ubv995) ([raw file](https://github.com/netsec-ethz/scion-coord/blob/d48537844251649b0c92634eb6c7b59b5fb2e099/models/as.go#L140)):*
> ```Go
> }
> 
> func (as *As) UpdateCurrency(Credits int64) error {
> ```

I woud separate all the virtual currency related code into a different library.

---

*[models/virtual_currency.go, line 1 at r4](https://reviewable.io:443/reviews/netsec-ethz/scion-coord/37#-Kva_fBuZsN5nxgI5_xl:-Kva_fBuZsN5nxgI5_xm:b8jw4p8) ([raw file](https://github.com/netsec-ethz/scion-coord/blob/d48537844251649b0c92634eb6c7b59b5fb2e099/models/virtual_currency.go#L1)):*
> ```Go
> package models
> ```

missing header

---

*[models/virtual_currency.go, line 5 at r4](https://reviewable.io:443/reviews/netsec-ethz/scion-coord/37#-Kva_hE27UztjvTrV-kT:-Kva_hE27UztjvTrV-kU:b-jh5t33) ([raw file](https://github.com/netsec-ethz/scion-coord/blob/d48537844251649b0c92634eb6c7b59b5fb2e099/models/virtual_currency.go#L5)):*
> ```Go
> 
> 
> func StartCredits() (int64){
> ```

I would comment each of the functions similar to the rest of the code-base. It also needs a bit of documentation why you used 100 as a starting credit.

---

*[models/virtual_currency.go, line 10 at r4](https://reviewable.io:443/reviews/netsec-ethz/scion-coord/37#-KvabVZDquE5O2VuhFK2:-KvabVZDquE5O2VuhFK3:b-g3v7tt) ([raw file](https://github.com/netsec-ethz/scion-coord/blob/d48537844251649b0c92634eb6c7b59b5fb2e099/models/virtual_currency.go#L10)):*
> ```Go
> 
> func BandwidthToCredits(bandwidthInKilobits uint64) (int64){
> 	return int64(bandwidthInKilobits) / 1000 / 10
> ```

I would just use `/100` instead of `/1000/10`.

---

*[models/virtual_currency.go, line 13 at r4](https://reviewable.io:443/reviews/netsec-ethz/scion-coord/37#-KvabkNPplNAqLxXpJQO:-KvabkNPplNAqLxXpJQP:brgbzyc) ([raw file](https://github.com/netsec-ethz/scion-coord/blob/d48537844251649b0c92634eb6c7b59b5fb2e099/models/virtual_currency.go#L13)):*
> ```Go
> }
> func CreditsToBandwidthInBandwidthInKilobits(credits uint64)(int64){
> 	return int64(credits) * 1000 * 10
> ```

Similar here.

---

*[models/virtual_currency.go, line 14 at r4](https://reviewable.io:443/reviews/netsec-ethz/scion-coord/37#-KvabQXzE0wkiKKl8Uzl:-KvabQY-QmhEybSOSj6_:bjovvub) ([raw file](https://github.com/netsec-ethz/scion-coord/blob/d48537844251649b0c92634eb6c7b59b5fb2e099/models/virtual_currency.go#L14)):*
> ```Go
> func CreditsToBandwidthInBandwidthInKilobits(credits uint64)(int64){
> 	return int64(credits) * 1000 * 10
> }
> ```

missing new line

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/netsec-ethz/scion-coord/37#-:-KvaalEfhdSzhdG2sJm4:b-3w2wzo)*
<!-- Sent from Reviewable.io -->

Meldanor commented:
Updated the PR:

Implemented the feedback you've given. There is now a possibility to switch on/off the system via a configuration file.

I cannot reproduce the error by CircleCI. I can start and use the coordinator and do not get the followning error (and no idea how to get them):
```bash
#!/bin/bash -eo pipefail
go test -v ./...

# github.com/netsec-ethz/scion-coord
./main.go:119: too many arguments in call to tollbooth.NewLimiter
	have (number, time.Duration, *limiter.ExpirableOptions)
	want (int64, *limiter.ExpirableOptions)
Exited with code 2
```

mlegner commented:
@Meldanor The CircleCI error will be fixed by PR #94.

Meldanor commented:
@mlegner Thank you for the information.

mlegner commented:
@Meldanor PR #94 is merged now; after rebasing the CircleCI error should vanish.

ercanucan commented:
Made a pass. @mlegner could you also have a look at this to see if I missed something?

---

Review status: 0 of 7 files reviewed at latest revision, 21 unresolved discussions, some commit checks failed.

---

*[main.go, line 229 at r6](https://reviewable.io:443/reviews/netsec-ethz/scion-coord/37#-KyLehhSdbeCayzC0CeE:-KyLehhTOgxXUpTEhiWE:bq4iprt) ([raw file](https://github.com/netsec-ethz/scion-coord/blob/5656a04265f1b7189f3f0161a449b38608690ec5/main.go#L229)):*
> ```Go
> 	// Virtual currency API
> 	// ==========================================================
> 	router.Handle("/api/listASConnections/{account_id}/{secret}/{isdas}", apiChain.ThenFunc(asController.ListAsesConnectionsWithCredits))
> ```

Please pay attention to line lengths. We typically avoid lines longer than 100 characters.

---

*[conf/development.conf.default, line 63 at r6](https://reviewable.io:443/reviews/netsec-ethz/scion-coord/37#-KyLeydHkTiFpo1o2xJ7:-KyLeydHkTiFpo1o2xJ8:bh5eyyo) ([raw file](https://github.com/netsec-ethz/scion-coord/blob/5656a04265f1b7189f3f0161a449b38608690ec5/conf/development.conf.default#L63)):*
> ```
> 
> # Virtual Credit System
> virtualCredit.enable = true
> ```

Can we make it disabled by default, as we do not yet have everything in place for this.

---

*[conf/development.conf.default, line 65 at r6](https://reviewable.io:443/reviews/netsec-ethz/scion-coord/37#-KyLesmrsqTTNMpCa6UX:-KyLesmrsqTTNMpCa6UY:bjovvub) ([raw file](https://github.com/netsec-ethz/scion-coord/blob/5656a04265f1b7189f3f0161a449b38608690ec5/conf/development.conf.default#L65)):*
> ```
> virtualCredit.enable = true
> # 1 Credit = 10 MegaBits/s worth of bandwidth
> virtualCredit.startCredit = 100
> ```

missing new line

---

*[controllers/api/virtual_credit.go, line 1 at r6](https://reviewable.io:443/reviews/netsec-ethz/scion-coord/37#-KyLgI949qTxpV9aqtBL:-KyLgI949qTxpV9aqtBM:bb6ht2h) ([raw file](https://github.com/netsec-ethz/scion-coord/blob/5656a04265f1b7189f3f0161a449b38608690ec5/controllers/api/virtual_credit.go#L1)):*
> ```Go
> // Copyright 2016 ETH Zurich
> ```

2017

---

*[controllers/api/virtual_credit.go, line 65 at r6](https://reviewable.io:443/reviews/netsec-ethz/scion-coord/37#-KyLhwwpa2Ye_DdGVS7e:-KyLhwwqUhU7ov5qRVPE:b-b0vlza) ([raw file](https://github.com/netsec-ethz/scion-coord/blob/5656a04265f1b7189f3f0161a449b38608690ec5/controllers/api/virtual_credit.go#L65)):*
> ```Go
> 
> 	vars := mux.Vars(r)
> 	isdas := vars["isdas"]
> ```

it may be better use the following syntax for looking up a key in a map:
```
isdas, ok := m["route"]
```

---

*[controllers/api/virtual_credit.go, line 120 at r6](https://reviewable.io:443/reviews/netsec-ethz/scion-coord/37#-KyLj-wR4YAgFxcBXqms:-KyLj-wR4YAgFxcBXqmt:b-lg8kjb) ([raw file](https://github.com/netsec-ethz/scion-coord/blob/5656a04265f1b7189f3f0161a449b38608690ec5/controllers/api/virtual_credit.go#L120)):*
> ```Go
> 	// Subtract credits from AS
> 	if err := as.UpdateCurrency(-1 * creditsNeeded); err != nil {
> 		log.Printf("Error: Substracting credits! AS: %v, Request: %v, Error: %v", as, r.Body, err)
> ```

Subtracting

---

*[controllers/api/virtual_credit.go, line 127 at r6](https://reviewable.io:443/reviews/netsec-ethz/scion-coord/37#-KyLjiSyAuXP3sUKg0WX:-KyLjiSyAuXP3sUKg0WY:b-1uz1z9) ([raw file](https://github.com/netsec-ethz/scion-coord/blob/5656a04265f1b7189f3f0161a449b38608690ec5/controllers/api/virtual_credit.go#L127)):*
> ```Go
> }
> 
> func (c *ASController) checkAndUpdateCreditsAtResponse(w http.ResponseWriter, r *http.Request, cr *models.ConnRequest, reply ConnReply) error {
> ```

I would add brief documentation to the top of each function

---

*[controllers/api/virtual_credit.go, line 174 at r6](https://reviewable.io:443/reviews/netsec-ethz/scion-coord/37#-KyLjxKdia9cK54czujV:-KyLjxKdia9cK54czujW:bmlpur7) ([raw file](https://github.com/netsec-ethz/scion-coord/blob/5656a04265f1b7189f3f0161a449b38608690ec5/controllers/api/virtual_credit.go#L174)):*
> ```Go
> 	// Roll back UpdateCurrency changes
> 	if err := as.UpdateCurrency(creditsNeeded); err != nil {
> 		log.Printf("Critical error: Can't roll back UpdateCurrency changes! Credits: %v, AS: %v, Request: %v, Error: %v", creditsNeeded, as, r.Body, err)
> ```

very long line

---

*[models/virtual_credit.go, line 1 at r6](https://reviewable.io:443/reviews/netsec-ethz/scion-coord/37#-KyLk0JOjriJo1xA3736:-KyLk0JOjriJo1xA3737:bb6ht2h) ([raw file](https://github.com/netsec-ethz/scion-coord/blob/5656a04265f1b7189f3f0161a449b38608690ec5/models/virtual_credit.go#L1)):*
> ```Go
> // Copyright 2016 ETH Zurich
> ```

2017

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/netsec-ethz/scion-coord/37#-:-KyLk5wfIaskvO_X01X9:b-iatks8)*
<!-- Sent from Reviewable.io -->

Meldanor commented:
Implemented your suggestions, thanks for that! Sorry, if something is clunky and the code isn't well polished. I'm new to Golang and very used to Java ^^

mlegner commented:
I also made a pass, mostly minor typos and coding-convention "violations".

---

Review status: 0 of 7 files reviewed at latest revision, 43 unresolved discussions, some commit checks failed.

---

*[main.go, line 228 at r7](https://reviewable.io:443/reviews/netsec-ethz/scion-coord/37#-Kypw_u1MenHx8AoLYdM:-Kypw_u2TjMKlDIXJbxx:bxv4scn) ([raw file](https://github.com/netsec-ethz/scion-coord/blob/0d7bbab686953e510760a3b9cbd7dbc4c0fe6ed5/main.go#L228)):*
> ```Go
> 
> 	// Virtual currency API
> 	// ==========================================================
> ```

You could keep the same layout as on other titles: "===" line above title.

---

*[config/config.go, line 61 at r7](https://reviewable.io:443/reviews/netsec-ethz/scion-coord/37#-KypxKQnN6h2aLaG6CaV:-KypxKQnN6h2aLaG6CaW:bywf1q1) ([raw file](https://github.com/netsec-ethz/scion-coord/blob/0d7bbab686953e510760a3b9cbd7dbc4c0fe6ed5/config/config.go#L61)):*
> ```Go
> 
> 	// Virtual Credit system
> 	// If enable = false -> every method will immatedly return nil or nothing
> ```

This explanation may be more helpful in the development.conf.default. Also there is a typo "immatedly".

---

*[config/config.go, line 62 at r7](https://reviewable.io:443/reviews/netsec-ethz/scion-coord/37#-KypxBQiqVNexoKw6vMZ:-KypxBQiqVNexoKw6vM_:b-sbmmbv) ([raw file](https://github.com/netsec-ethz/scion-coord/blob/0d7bbab686953e510760a3b9cbd7dbc4c0fe6ed5/config/config.go#L62)):*
> ```Go
> 	// Virtual Credit system
> 	// If enable = false -> every method will immatedly return nil or nothing
> 	VIRTUAL_CREDIT_ENABLE, _        = goconf.AppConf.Bool("virtualCredit.enable")
> ```

You could move these settings above the "VM_LOCAL_IP".

---

*[config/config.go, line 63 at r7](https://reviewable.io:443/reviews/netsec-ethz/scion-coord/37#-KypxVqq12dmqRCMxxBv:-KypxVqq12dmqRCMxxBw:bmmoa2z) ([raw file](https://github.com/netsec-ethz/scion-coord/blob/0d7bbab686953e510760a3b9cbd7dbc4c0fe6ed5/config/config.go#L63)):*
> ```Go
> 	// If enable = false -> every method will immatedly return nil or nothing
> 	VIRTUAL_CREDIT_ENABLE, _        = goconf.AppConf.Bool("virtualCredit.enable")
> 	VIRTUAL_CREDIT_START_CREDITS, _ = goconf.AppConf.Int64("virtualCredit.startCredits")
> ```

Does this have to be an "Int64"?

---

*[controllers/api/as.go, line 341 at r7](https://reviewable.io:443/reviews/netsec-ethz/scion-coord/37#-KypybZCXYlGx0hre5Zf:-KypybZCXYlGx0hre5Zg:b93lyfk) ([raw file](https://github.com/netsec-ethz/scion-coord/blob/0d7bbab686953e510760a3b9cbd7dbc4c0fe6ed5/controllers/api/as.go#L341)):*
> ```Go
> 	// update new credits for both ASes
> 	if err := c.checkAndUpdateCredits(w, r, &cr); err != nil {
> 		// logs the error and writes back the response
> ```

You could be more clear in saying that the function checkAndUpdateCredits takes care of logging and writing the response.

---

*[controllers/api/as.go, line 366 at r7](https://reviewable.io:443/reviews/netsec-ethz/scion-coord/37#-Kypz-1Ww0uHRpQNFWNV:-Kypz-1XiNmfnYo-3knH:b-v13kbz) ([raw file](https://github.com/netsec-ethz/scion-coord/blob/0d7bbab686953e510760a3b9cbd7dbc4c0fe6ed5/controllers/api/as.go#L366)):*
> ```Go
> 			cr.RequestIA, err)
> 		c.Error500(err, w, r)
> 		// The credits will be granted in hindsight and must be removed in case of an error (now)
> ```

Is "hindsight" the correct word here?

---

*[controllers/api/as.go, line 430 at r7](https://reviewable.io:443/reviews/netsec-ethz/scion-coord/37#-Kypz5beMDVmFwQMQLSx:-Kypz5beMDVmFwQMQLSy:bmr1ldi) ([raw file](https://github.com/netsec-ethz/scion-coord/blob/0d7bbab686953e510760a3b9cbd7dbc4c0fe6ed5/controllers/api/as.go#L430)):*
> ```Go
> 	// DENIED = give credits back to the initiator
> 	if err := c.checkAndUpdateCreditsAtResponse(w, r, cr, reply); err != nil {
> 		// logs the error and writes back the response
> ```

Same as above, make it more clear that the function takes care of that.

---

*[controllers/api/virtual_credit.go, line 120 at r6](https://reviewable.io:443/reviews/netsec-ethz/scion-coord/37#-KyLj-wR4YAgFxcBXqms:-Kyq5Nd4TtsLxSW_2rgc:bvtk1rg) ([raw file](https://github.com/netsec-ethz/scion-coord/blob/5656a04265f1b7189f3f0161a449b38608690ec5/controllers/api/virtual_credit.go#L120)):*
<details><summary><i>Previously, ercanucan (ercanucan) wrote…</i></summary><blockquote>

Subtracting
</blockquote></details>

Not yet resolved.

---

*[controllers/api/virtual_credit.go, line 34 at r7](https://reviewable.io:443/reviews/netsec-ethz/scion-coord/37#-KypzSUY5xtWjGemRX4Q:-KypzSUY5xtWjGemRX4R:b6zowzj) ([raw file](https://github.com/netsec-ethz/scion-coord/blob/0d7bbab686953e510760a3b9cbd7dbc4c0fe6ed5/controllers/api/virtual_credit.go#L34)):*
> ```Go
> // REST resource to list the payed connections from one AS (identified by the parameter 'isdas')
> // Example Response:
> /*
> ```

We don't normally use multiline comments.

---

*[controllers/api/virtual_credit.go, line 51 at r7](https://reviewable.io:443/reviews/netsec-ethz/scion-coord/37#-Kyq-3VFEzKSERZHkaRZ:-Kyq-3VFEzKSERZHkaR_:bg7tyn9) ([raw file](https://github.com/netsec-ethz/scion-coord/blob/0d7bbab686953e510760a3b9cbd7dbc4c0fe6ed5/controllers/api/virtual_credit.go#L51)):*
> ```Go
> }
> */
> func (c *ASController) ListAsesConnectionsWithCredits(w http.ResponseWriter, r *http.Request) {
> ```

I would use "ASes" instead of "Ases". Generally, have all letters of an abbreviation as either uppercase or  lowercase.

---

*[controllers/api/virtual_credit.go, line 60 at r7](https://reviewable.io:443/reviews/netsec-ethz/scion-coord/37#-Kyq-U-1pfOlrb1cFQVJ:-Kyq-U-1pfOlrb1cFQVK:b-espfij) ([raw file](https://github.com/netsec-ethz/scion-coord/blob/0d7bbab686953e510760a3b9cbd7dbc4c0fe6ed5/controllers/api/virtual_credit.go#L60)):*
> ```Go
> 		ISD         int                            // The ISD of this AS
> 		AS          int                            // This AS
> 		Credits     int64                          // The current credits of this AS
> ```

Does this have to be an int64?

---

*[controllers/api/virtual_credit.go, line 73 at r7](https://reviewable.io:443/reviews/netsec-ethz/scion-coord/37#-Kyq-cdpb-YyBANDmske:-Kyq-cdpb-YyBANDmskf:b-jes3u6) ([raw file](https://github.com/netsec-ethz/scion-coord/blob/0d7bbab686953e510760a3b9cbd7dbc4c0fe6ed5/controllers/api/virtual_credit.go#L73)):*
> ```Go
> 	requestingAS, err := models.FindAsByIsdAs(isdas)
> 	if err != nil {
> 		c.NotFound(errors.New(isdas+" not found"), w, r)
> ```

You could use `fmt.Errorf("%v not found", isdas)` instead of `errors.New`.

---

*[controllers/api/virtual_credit.go, line 97 at r7](https://reviewable.io:443/reviews/netsec-ethz/scion-coord/37#-Kypzbjk4Z6mdfmR0lJT:-Kypzbjk4Z6mdfmR0lJU:b-3s9bum) ([raw file](https://github.com/netsec-ethz/scion-coord/blob/0d7bbab686953e510760a3b9cbd7dbc4c0fe6ed5/controllers/api/virtual_credit.go#L97)):*
> ```Go
> }
> 
> /*
> ```

Multiline comments.

---

*[controllers/api/virtual_credit.go, line 137 at r7](https://reviewable.io:443/reviews/netsec-ethz/scion-coord/37#-Kypzed27TVazrGJVvbN:-Kypzed27TVazrGJVvbO:b-3s9bum) ([raw file](https://github.com/netsec-ethz/scion-coord/blob/0d7bbab686953e510760a3b9cbd7dbc4c0fe6ed5/controllers/api/virtual_credit.go#L137)):*
> ```Go
> }
> 
> /*
> ```

Multiline comments.

---

*[controllers/api/virtual_credit.go, line 138 at r7](https://reviewable.io:443/reviews/netsec-ethz/scion-coord/37#-Kyq2UYjlJD5sabd8iCv:-Kyq2UYjlJD5sabd8iCw:bzc7ia9) ([raw file](https://github.com/netsec-ethz/scion-coord/blob/0d7bbab686953e510760a3b9cbd7dbc4c0fe6ed5/controllers/api/virtual_credit.go#L138)):*
> ```Go
> 
> /*
> 	Handles the credit transaction for the requester and the target when an ConnectionResponse is
> ```

*a* ConnectionResponse

---

*[controllers/api/virtual_credit.go, line 152 at r7](https://reviewable.io:443/reviews/netsec-ethz/scion-coord/37#-Kyq4Yi04S2tX6JiHSgc:-Kyq4Yi04S2tX6JiHSgd:bwnmy80) ([raw file](https://github.com/netsec-ethz/scion-coord/blob/0d7bbab686953e510760a3b9cbd7dbc4c0fe6ed5/controllers/api/virtual_credit.go#L152)):*
> ```Go
> 
> 	as, _ := models.FindAsByIsdAs(cr.RequestIA)
> 	var credits = models.BandwidthToCredits(cr.Bandwidth)
> ```

We usually use the `:=` syntax instead of `var`.

---

*[controllers/api/virtual_credit.go, line 178 at r7](https://reviewable.io:443/reviews/netsec-ethz/scion-coord/37#-KypzfSu8ogqSkkKcD-4:-KypzfSu8ogqSkkKcD-5:b-3s9bum) ([raw file](https://github.com/netsec-ethz/scion-coord/blob/0d7bbab686953e510760a3b9cbd7dbc4c0fe6ed5/controllers/api/virtual_credit.go#L178)):*
> ```Go
> }
> 
> /*
> ```

Multiline comments.

---

*[controllers/api/virtual_credit.go, line 179 at r7](https://reviewable.io:443/reviews/netsec-ethz/scion-coord/37#-Kyq2ZuKY-jZtDaII4jg:-Kyq2ZuKY-jZtDaII4jh:b-mjya55) ([raw file](https://github.com/netsec-ethz/scion-coord/blob/0d7bbab686953e510760a3b9cbd7dbc4c0fe6ed5/controllers/api/virtual_credit.go#L179)):*
> ```Go
> 
> /*
> 	If an error occurs while handling the COnnectionResponse, this function will rollback the change
> ```

"COnnection"

---

*[controllers/api/virtual_credit.go, line 196 at r7](https://reviewable.io:443/reviews/netsec-ethz/scion-coord/37#-Kyq4gp3ppDwJJ5j1Q8F:-Kyq4gp3ppDwJJ5j1Q8G:bwnmy80) ([raw file](https://github.com/netsec-ethz/scion-coord/blob/0d7bbab686953e510760a3b9cbd7dbc4c0fe6ed5/controllers/api/virtual_credit.go#L196)):*
> ```Go
> 	}
> 
> 	var creditsNeeded = models.BandwidthToCredits(cr.Bandwidth)
> ```

We usually use the `:=` syntax instead of `var`.

---

*[models/as.go, line 140 at r4](https://reviewable.io:443/reviews/netsec-ethz/scion-coord/37#-Kva_phdbAHixoGtf4ap:-Kyq3lH5UWMJ3tbQwffZ:b-gycalv) ([raw file](https://github.com/netsec-ethz/scion-coord/blob/d48537844251649b0c92634eb6c7b59b5fb2e099/models/as.go#L140)):*
<details><summary><i>Previously, ercanucan (ercanucan) wrote…</i></summary><blockquote>

I woud separate all the virtual currency related code into a different library.
</blockquote></details>

I agree, you could also move this function to "models/virtual_credit.go".

---

*[models/as.go, line 1 at r7](https://reviewable.io:443/reviews/netsec-ethz/scion-coord/37#-Kyq171ulVxB2dQz2X6J:-Kyq171ulVxB2dQz2X6K:b-58slew) ([raw file](https://github.com/netsec-ethz/scion-coord/blob/0d7bbab686953e510760a3b9cbd7dbc4c0fe6ed5/models/as.go#L1)):*
> ```Go
> // Copyright 2017 ETH Zurich
> ```

We generally keep the year where the file was first created.

---

*[models/as.go, line 33 at r7](https://reviewable.io:443/reviews/netsec-ethz/scion-coord/37#-Kyq1Ay85YqZaHLi0jNU:-Kyq1Ay85YqZaHLi0jNV:b-eskida) ([raw file](https://github.com/netsec-ethz/scion-coord/blob/0d7bbab686953e510760a3b9cbd7dbc4c0fe6ed5/models/as.go#L33)):*
> ```Go
> 	Core    bool     `orm:"default(false)"`
> 	Account *Account `orm:"rel(fk);index"`
> 	Credits int64
> ```

Does this have to be int64?

---

*[models/virtual_credit.go, line 1 at r6](https://reviewable.io:443/reviews/netsec-ethz/scion-coord/37#-KyLk0JOjriJo1xA3736:-Kyq3xEnPWczfHwNH5iD:bvwq4zk) ([raw file](https://github.com/netsec-ethz/scion-coord/blob/5656a04265f1b7189f3f0161a449b38608690ec5/models/virtual_credit.go#L1)):*
<details><summary><i>Previously, ercanucan (ercanucan) wrote…</i></summary><blockquote>

2017
</blockquote></details>

Still unresolved.

---

*[models/virtual_credit.go, line 25 at r7](https://reviewable.io:443/reviews/netsec-ethz/scion-coord/37#-Kyq3zyrFGQbMBSDkbmg:-Kyq3zyrFGQbMBSDkbmh:b-vbt9t7) ([raw file](https://github.com/netsec-ethz/scion-coord/blob/0d7bbab686953e510760a3b9cbd7dbc4c0fe6ed5/models/virtual_credit.go#L25)):*
> ```Go
> 
> // The start credits a new AS gets
> func StartCredits() int64 {
> ```

Do all of those need to be int64?
Also, is it necessary to have a function simply returning the defined constant?

---

*[models/virtual_credit.go, line 51 at r7](https://reviewable.io:443/reviews/netsec-ethz/scion-coord/37#-Kyq4j5QdSBvX9PuVvdU:-Kyq4j5QdSBvX9PuVvdV:bi9f3bx) ([raw file](https://github.com/netsec-ethz/scion-coord/blob/0d7bbab686953e510760a3b9cbd7dbc4c0fe6ed5/models/virtual_credit.go#L51)):*
> ```Go
> // Look for all connection to and from this AS and calculates the necessary credits for it
> func (as *As) ListConnections() ([]ConnectionWithCredits, error) {
> 	var connections []ConnectionWithCredits
> ```

You may want to replace most occurences of `var` with the equivalent `:=`.

---

*[models/virtual_credit.go, line 72 at r7](https://reviewable.io:443/reviews/netsec-ethz/scion-coord/37#-Kyq4voFkt9nn7mZvPcE:-Kyq4voGozg9ewuSKdaU:b-w8f92t) ([raw file](https://github.com/netsec-ethz/scion-coord/blob/0d7bbab686953e510760a3b9cbd7dbc4c0fe6ed5/models/virtual_credit.go#L72)):*
> ```Go
> 	}
> 
> 	// incoming ones (this AS get Credits for)
> ```

Incoming

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/netsec-ethz/scion-coord/37#-:-Kyq5A6j6SHFCbMIfZMa:b2v7ch9)*
<!-- Sent from Reviewable.io -->

Meldanor commented:
Implemented (hopefully) your suggestion, but one.

I want to use `int64` instead of `int` to store the Credits, I do not see any advantage to only use 32 bit long integers when we are producing a long term system like SCION.

mlegner commented:
I see your point that it doesn't really matter in terms of space. However, I also don't see a real advantage of 64bit integers. In particular, when starting credits are 100, 2^31 > 2*10^9 should suffice overall, don't you think? Also, if at some point, we realize that 32bit are too short, we won't have any problems to change it to 64bit with an according change in the database. However, if you feel strongly about this, I won't stand in the way.
I resolved most of my comments, some are still valid.
@ercanucan: Could you resolve your comments as well?

---

Review status: 0 of 7 files reviewed at latest revision, 21 unresolved discussions.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/netsec-ethz/scion-coord/37#-:-Kyuo6t1zOVEv84CUDgB:bgmrsn3)*
<!-- Sent from Reviewable.io -->

ercanucan commented:
Reviewed 1 of 8 files at r5, 7 of 7 files at r8.
Review status: all files reviewed at latest revision, 22 unresolved discussions.

---

*[controllers/api/as.go, line 426 at r7](https://reviewable.io:443/reviews/netsec-ethz/scion-coord/37#-KyaipT4vxpPEgqyusT0:-KyaipT5QBVgTIZBENnL:b-dcwyc6) ([raw file](https://github.com/netsec-ethz/scion-coord/blob/0d7bbab686953e510760a3b9cbd7dbc4c0fe6ed5/controllers/api/as.go#L426)):*
> ```Go
> 	}
> 
> 	// Add / Substract the credits for the request depending on the ConnReply.Status
> ```

minor: Subtract

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/netsec-ethz/scion-coord/37)*
<!-- Sent from Reviewable.io -->

Meldanor commented:
Review status: 4 of 7 files reviewed at latest revision, 22 unresolved discussions, some commit checks failed.

---

*[controllers/api/virtual_credit.go, line 120 at r6](https://reviewable.io:443/reviews/netsec-ethz/scion-coord/37#-KyLj-wR4YAgFxcBXqms:-KzeZPo8ArUGRPkWB-UZ:baasjez) ([raw file](https://github.com/netsec-ethz/scion-coord/blob/5656a04265f1b7189f3f0161a449b38608690ec5/controllers/api/virtual_credit.go#L120)):*
<details><summary><i>Previously, mlegner (Markus Legner) wrote…</i></summary><blockquote>

Not yet resolved.
</blockquote></details>

I looked at the comment and though "? Why is this not resolved?". Yeah, ok, the error was in Printf...

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/netsec-ethz/scion-coord/37)*
<!-- Sent from Reviewable.io -->

Meldanor commented:
I hope, I processed every issue. I'm not used to Reviewable and its type of presentation.

hausheer commented:
Thank you Kilian, Markus, I see that Github states that the reviewable check is still pending and that the branch would need to be merged again with the latest changes from master. Could you take another look at it please? It would be great if you could resolve this soon, to avoid that the PR gets out-of-date again.

Meldanor commented:
How do you handle "merging"? This PR has many many commits which do only make sense when you know its origin , this PR. So do I need to make a rebase and squash them all into one commit?

hausheer commented:
@mlegner is a rebase sufficient or does @Meldanor need to squash the commits into one as well?

mlegner commented:
@meldanor Made a pass and resolved all my remaining issues.
We generally both rebase and squash commits before merging them into master in order to keep a nice linear history. So yes, please first squash them into one commit and then rebase it onto the current master (it is easier this way in case there are conflicts).
@ercanucan Could you resolve your remaining discussions?

---

Review status: 4 of 7 files reviewed at latest revision, 21 unresolved discussions, some commit checks failed.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/netsec-ethz/scion-coord/37#-:-Kzirsfku8ffXqjL2jGw:b-pxiecn)*
<!-- Sent from Reviewable.io -->

Meldanor commented:
Ok, I will wait with the Rebase when everything is resolved.

ercanucan commented:
Reviewed 3 of 3 files at r9.
Review status: all files reviewed at latest revision, 21 unresolved discussions, some commit checks failed.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/netsec-ethz/scion-coord/37)*
<!-- Sent from Reviewable.io -->

ercanucan commented:
@mlegner I marked the discussions resolved.
@meldanor Please squash the commits into one. You can do this by using `rebase -i` command.

---

Review status: all files reviewed at latest revision, 21 unresolved discussions, some commit checks failed.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/netsec-ethz/scion-coord/37#-:-KzjXNFRjHtPH8FrHKW6:br4suew)*
<!-- Sent from Reviewable.io -->

Meldanor commented:
I squashed the commit into one and the PR is now mergeable.

ercanucan commented:
Reviewed 3 of 3 files at r10.
Review status: all files reviewed at latest revision, 21 unresolved discussions.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/netsec-ethz/scion-coord/37)*
<!-- Sent from Reviewable.io -->

mlegner commented:
@ercanucan The issues are still not resolved for me. Could you have another look?

---

Review status: all files reviewed at latest revision, 21 unresolved discussions.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/netsec-ethz/scion-coord/37#-:-KznbwmGsOSoINIo2DIV:b-3n51dj)*
<!-- Sent from Reviewable.io -->

ercanucan commented:
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

---

Review status: all files reviewed at latest revision, all discussions resolved, some commit checks failed.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/netsec-ethz/scion-coord/37#-:-KznjxsAlAYBr9oS1NR_:bnfp4nl)*
<!-- Sent from Reviewable.io -->

ercanucan commented:
@mlegner should be done now.

---

Review status: all files reviewed at latest revision, all discussions resolved, some commit checks failed.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/netsec-ethz/scion-coord/37#-:-Kznk-SCVsXJ7RTz8nhA:b-nf76bf)*
<!-- Sent from Reviewable.io -->

mlegner commented:
Thanks for resolving the discussions.
@Meldanor Could you rebase it on the current master branch, then we can merge this PR.

Meldanor commented:
Rebased on current master branch.
